### ISSUE TITLE
support generic lambda.HandlerFunc for otellambda

### ISSUE THREAD

**Issue by jan-xyz** [open-telemetry/opentelemetry-go-contrib#3660]
The Go lambda framework now has support for a generic signature that verifies that the handler has the correct signature. The new functions are

```go
// function type
lambda.StartHandlerFunc

// struct type
lambda.StartHandler
lambda.StartHandlerWithContext
```

that doesn't work anymore with the `otellambda.InstrumentHandler`. Can we add support for the generic types? With a bit of direction and a thumbs up, I could also contribute this.

**Comment by rackerWard** [comment]
How would you implement in Go an API to accept the different handler function declaration signatures without creating a uniquely named function for each kind?  

Lambda handlers can be of the following formats.

```
func()
func(context)
func(context, event)
func(event)

func() error
func(context) error
func(context, event) error
func(event) error

func() (response, error)
func(context) (response, error
func(context, event) (response, error)
func(event) (response, error)
```

**Comment by jan-xyz** [comment]
I don't think it is necessary to support all permutations and only provide an additional function that can deal with the type that is also used by [StartHandlerFunc](https://github.com/aws/aws-lambda-go/blob/v1.41.0/lambda/entry_generic.go#L19):

```go
// HandlerFunc represents a valid input with two arguments and two returns as described by Start
type HandlerFunc[TIn, TOut any] interface {
	func(context.Context, TIn) (TOut, error)
}
```

This limits it basically to `func(context, event) (response, error)`, which is the most common use-case probably. All other options are a subset of this anyways and can be implemented with functions that just don't use the parameters and return values that the interface provides.

**Comment by lukasmalkmus** [comment]
@jan-xyz Have you made any progress on this?

Seeing https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5545 and https://github.com/open-telemetry/opentelemetry-go-contrib/issues/5546, I guess, no?

I was initially very happy to stumble onto those packages. Some rough edges here and there - like those issue here - but overall it works just fine. Sad to see those two packages being deprecated. 

Not sure if I'm up to becoming an owner just yet, tho.
